// nodejs-nocaptcha-recaptcha/index.js
/**
* A module to verify Google's new no catpcha recaptcha
*
* @param {string}, @param {string}, @param {callback}
* @return {boleean}
*/

var debug = require('debug')("recaptcha");
var https = require('https');

module.exports = (
	function (response, secret, callback) {

		var data='';
		var post_data = "secret=" + secret + "&response=" + response; 
		var httpObject = {     
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded',
				'Content-Length': post_data.length
			},
			hostname: "www.google.com",
			port: 443,
			path: "/recaptcha/api/siteverify",
			method: 'POST',
			rejectUnauthorized: true,
		};
		var req = https.request( httpObject, function (res){
			res.setEncoding('utf8');
			
			res.on('data', function (chunk){
				data += chunk;
			});
			res.on('end', function () {
				debug(data);
				var result = JSON.parse(data)
				callback(JSON.parse(result.success));
			});
			req.on('error', function (e) {
				debug("Got error: " + e.message);
				callback(false);
			});
			req.on('upgrade', function (e){
				debug(e);
			});
		});	

		req.write(post_data,'utf8');
		req.end();

	}
);
